Add slides on modules

bernhardmgruber opened the pull request:
This includes:

* motivation for modules
* writing modules
* export
* visibility/reachability
* module structure
* module interface partitions
* module implementation units
* header units
* include translation
* build workflow
* build system challenges
* build system status
* g++ case study

Fixes: #53

hageboeck commented on talk/expert/modules.tex:
```diff
@@ -0,0 +1,510 @@
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Modules}
+  \begin{block}{Motivation}
+    \begin{itemize}
+    \item Better isolation
+    \begin{itemize}
+      \item Between multiple pieces of included code
+      \item Between included and your code
+    \end{itemize}
+    \item Better control over public interface of your library
+    \begin{itemize}
+      \item Avoid leaking library dependencies into user code
+    \end{itemize}
+    \item Faster compilation
+    \end{itemize}
```
As the very first thing, I would have liked to see what a module is. Here's a try:
```suggestion
    \begin{itemize}
      \item Modules allow for sharing declarations and definitions across translation units
      \item Header files do the same, but modules have benefits:
      \begin{itemize}
        \item Better isolation. No cross-talk between multiple pieces of included code, and also not between included code an your code.
        \item Better control over public interface of your library. Avoid leaking library dependencies into user code.
        \item Faster compilation
      \end{itemize}
    \end{itemize}
```

bernhardmgruber replied:
Not bad, I take that.

hageboeck commented on talk/expert/modules.tex:
```diff
@@ -0,0 +1,510 @@
+    \end{block}
+    \begin{block}{Include translation}
+      \begin{itemize}
+        \item Allows the compiler to automatically turn includes into imports.
+        \item Controlled via compiler flags and the build system.
+        \item Basically a standard replacement for precompiled headers.
```
... and so on.

hageboeck commented on talk/expert/modules.tex:
```diff
@@ -0,0 +1,510 @@
+    \frametitlecpp[20]{Building modules (g++)}
+    \begin{block}{Case study: g++}
+      \begin{itemize}
+        \item By default, g++ caches compiled module interfaces (CMI, i.e.\ BMI) in subdirectory \texttt{./gcm.cache}
```
Bravo!

bernhardmgruber replied:
Because I put the backslash in `i.e.\` myself? :D

hageboeck commented on talk/expert/modules.tex:
```diff
@@ -0,0 +1,510 @@
+        \item Each header unit needs to be built before it can be imported
+        \begin{itemize}
+          \item {\footnotesize \mintinline{cpp}{g++ -std=c++20 -fmodules-ts -x c++-system-header vector}}
+          \item Generates e.g. \texttt{./gcm.cache/usr/include/c++/11/vector.gcm} (CMI)
+          \item {\footnotesize \mintinline{cpp}{g++ -std=c++20 -fmodules-ts -x c++-header ratio.h}}
+          \item Generates e.g. \texttt{./gcm.cache/,/ratio.h.gcm} (CMI)
```
```suggestion
          \item Generates e.g.\ \texttt{./gcm.cache/usr/include/c++/11/vector.gcm} (CMI)
          \item {\footnotesize \mintinline{cpp}{g++ -std=c++20 -fmodules-ts -x c++-header ratio.h}}
          \item Generates e.g.\ \texttt{./gcm.cache/,/ratio.h.gcm} (CMI)
```

hageboeck reviewed (commented):
Here's a first batch of suggestions.
I'm not sure on the length of the "module" and the ordering of slides, though. Talk about this next week?

bernhardmgruber commented on talk/expert/modules.tex:
```diff
@@ -0,0 +1,510 @@
+    \frametitlecpp[20]{Header units}
+    \begin{block}{Purpose}
+      \begin{itemize}
+      \item For easier transition from header-based projects to modules.
+      \item Allows to import any header file as if it was a module.
+      \item Can be mixed with regular header inclusion.
```
Thx, I was always wondering what the rule is!

bernhardmgruber commented on talk/expert/modules.tex:
```diff
@@ -0,0 +1,510 @@
+    \frametitlecpp[20]{Header units}
+    \begin{block}{Effect}
+      \begin{itemize}
+      \item All declarations in the imported header will be exported.
+      \item An imported header is not affected by the preprocessor state.
+            Thus, you cannot "configure" the header by defining a macro before importing.
+            This allows the header to be precompiled.
+      \item An imported header may still provide preprocessor macros.
```
Thx!

bernhardmgruber commented on talk/expert/modules.tex:
```diff
@@ -0,0 +1,510 @@
+\end{frame}
+
+\begin{frame}[fragile]
+    \frametitlecpp[20]{Include translation}
```
ok!

bernhardmgruber commented on talk/expert/modules.tex:
```diff
@@ -0,0 +1,510 @@
+      \item You cannot export aliases to non-exported types
+    \end{itemize}
+  \end{block}
+\end{frame}
```
I think it's fine here. Let's discuss this later!

bernhardmgruber replied:
I reordered now.

bernhardmgruber commented:
@hageboeck I applied your feedback in person:

* I moved header units and include translation to the back
* I added two figures, one showing the traditional compilation workflow and one showing a module based one

hageboeck commented on talk/expert/modules.tex:
```diff
@@ -0,0 +1,643 @@
+      \item Modules allow for sharing declarations and definitions across translation units
+      \item Header files do the same, but modules have benefits:
+      \begin{itemize}
+        \item Better isolation. No cross-talk between multiple pieces of included code, and also not between included code an your code.
```
```suggestion
        \item Better isolation. No cross-talk between multiple pieces of included code, and also not between included code and your code.
```

hageboeck commented on talk/expert/modules.tex:
```diff
@@ -0,0 +1,643 @@
+    \begin{itemize}
+      \item By default, everything inside a module is internal
+      \item Declarations become visible when declared with \mintinline{cpp}{export} in the primary module interface
+      \item You can only export entities at namespace scope
```
I had to look up whether the global scope was allowed or not. Maybe somebody else would have had the same question? In that case, we could add:
```suggestion
      \item You can only export entities at namespace scope (including the global scope)
```

bernhardmgruber replied:
The global scope is a namespace scope, but I can see how this may be confusing.

hageboeck commented on talk/expert/modules.tex:
```diff
@@ -0,0 +1,643 @@
+    \begin{column}{.45\textwidth}
+      \begin{exampleblock}{Module bar}
+        \begin{cppcode*}{gobble=6}
+          export module hi;
```
Is this a typo?
```suggestion
          export module bar;
```

hageboeck commented on talk/expert/modules.tex:
```diff
@@ -0,0 +1,643 @@
+        #include <cassert> // for assert()
+        #include <climits> // for e.g. CHAR_BIT
+      \end{cppcode*}
+      \item Further modularization of the standard library is in progress
```
That's a contradiction to "for the entire C++ standard library" if it's still in progress.

bernhardmgruber replied:
Then I need to rephrase. There is progress on providing more modules of the standard library. So far only `std` and `std.compat` have been standardized. We may see more modules like: `containers`, `fundamentals`, 'network' etc. in the future.

bernhardmgruber replied:
It now reads: "Definition of further, smaller modules of the standard library is in progress (e.g. fundamentals, containers, networking, etc.)"

hageboeck commented on talk/expert/modules.tex:
```diff
@@ -0,0 +1,643 @@
+        \begin{itemize}
+          \item Extra tool needs to infer dependency graph before building
+          \item This graph needs to be communicated to the build system
+          \item We likely need a \href{https://wg21.link/P1689}{standard dependency file format}
```
```suggestion
          \item We will likely need a \href{https://wg21.link/P1689}{standard dependency file format}
```

hageboeck reviewed (commented):
This is looking good, but I have some fiddling with the figure in mind. Cannot do this with gh suggestions, though, so let me check out the branch. In the mean time, I found some last small things.

hageboeck commented:
@bernhardmgruber I pushed the commit to fine tune the pictures. Here are the screenshots:
![image](https://user-images.githubusercontent.com/16205615/184663115-2baded43-ee27-472d-bc48-c31351dfb360.png)

![image](https://user-images.githubusercontent.com/16205615/184663166-3611087a-b5b7-4d79-962d-f7e0b7963067.png)

If you want to change other things, I can re-push, but I would say we could merge if the CI agrees.

bernhardmgruber commented:
Great work @hageboeck, this looks fabulous!
